common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	weights = make(map[Limit]HostWeight, len(limits))
+	usedRPS = make(map[Limit]PerSecond, len(limits))
+	snap := a.snapshot()
+	for _, lim := range limits {
+		hosts, used := a.getWeightsLocked(lim, snap)
+		if len(hosts) > 0 {
+			usedRPS[lim] = used // limit is known, has some usage
+			if weight, ok := hosts[host]; ok {
+				weights[lim] = weight // host has a known weight
+			}
+		}
+	}
+	return weights, usedRPS
+}
+
+func (a *impl) GC() Metrics {
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	m := Metrics{}
+	snap := a.snapshot()
+	// TODO: too costly? can check the first-N% each time and it'll eventually visit all keys, demonstrated in tests.
+	for lim, dat := range a.usage {
+		for host, hist := range dat {
+			if _, gc := missedUpdateScalar(snap.now.Sub(hist.lastUpdate), snap); gc {
+				// clean up stale host data within limits
+				delete(dat, host)
+				m.RemovedHostLimits++
+			} else {
+				m.HostLimits++
+			}
+		}
+
+		// clean up stale limits
+		if len(dat) == 0 {
+			delete(a.usage, lim)
+			m.RemovedLimits++
+		} else {
+			m.Limits++
+		}
+	}
+
+	return m
+}
+
+// returns a snapshot of config and "now" for easier chaining through calls, and reducing calls to
+// non-trivial field types like dynamic config.
+func (a *impl) snapshot() configSnapshot {
+	return configSnapshot{
+		now:       a.clock.Now(),
+		weight:    a.cfg.NewDataWeight(),
+		rate:      a.cfg.UpdateRate(),
+		maxMissed: a.cfg.MaxMissedUpdates(),
+		gcAfter:   a.cfg.GcAfter(),
+	}
+}
+
+// missedUpdateScalar returns an amount to multiply old RPS data by, to account for missed updates outside SLA.
+func missedUpdateScalar(elapsed time.Duration, cfg configSnapshot) (scalar PerSecond, gc bool) {
+	reduce := PerSecond(1.0)
+
+	// fast path: check the bounds for "new enough to ignore"
+	if elapsed <= time.Duration(float64(cfg.rate)*cfg.maxMissed) {
+		return reduce, false // within SLA, no changes
+	}
+	// fast path: check the bounds for "old enough to prune"
+	if elapsed >= cfg.rate*time.Duration(cfg.gcAfter) {
+		return reduce, true
+	}
+
+	// slow path: account for missed updates by simulating 0-value updates.
+
+	// calculate missed updates, and compute an exponential decay as if we had kept track along the way
+	missed := float64(elapsed) / float64(cfg.rate) // how many missed updates (fractional, 1 == one full duration passed, 2.5 == 2.5 durations passed, etc)
+	if missed >= cfg.maxMissed {
+		// missed at least one update period beyond decay-after, calculate an exponential decay to the old values.
+		// floor to an int when doing so because:
+		// - precision isn't important
+		// - tests are a bit easier (more stable / less crazy-looking values)
+		// - as a bonus freebie: integer exponents are typically faster to compute
+		reduce = PerSecond(math.Pow(cfg.weight, math.Floor(missed)))
+	}
+
+	return reduce, false
+}
+
+func weighted[T numeric](newer, older T, weight float64) T {
```

Comment:
is this used?

Reply:
it is, yes:
```go
next = history{
	lastUpdate: snap.now,
	accepted:   weighted(aps, prev.accepted*reduce, snap.weight),
	rejected:   weighted(rps, prev.rejected*reduce, snap.weight),
}
```
the extra types in `numeric` are largely to simplify tests, I only use `PerSecond` in this file.